### PR TITLE
[EVM-Equivalence-YUL] Optimize opcodes

### PR DESCRIPTION
**Pull request by vladimirradosavljevic**
# What ❔
This PR optimizes opcodes by:
1. Moving loop processing in a function and marking it as noinline. This removes a lot of unneeded spills and reloads generated by the compiler.
2. Combining checks for pop and push. LLVM will optimize these checks into a single comparison.
3. Adding always inline to warmAddress address function. This forces LLVM to inline warmAddress function in all places.

## Checklist

- [x] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [ ] Tests for the changes have been added / updated.
- [ ] Documentation comments have been added / updated.

### CODE REVIEW

**Comment by vladimirradosavljevic** [comment]
> > @jrchatruc PTAL. Improvements can be seen [here](https://github.com/matter-labs/era-compiler-tester/pull/62) (please note that this PR contains one pending LLVM optimization that is mostly related to ADD opcode, so we can get 44.750 instead of 46.750).
> 
> This looks really good, wouldn't it be worth it to also force the no inline on the constructor `simulate` function?

Currently in the constructor `simulate` function that is inlined, LLVM doesn't generate unneeded spills and reloads. In case we see that this happens in a future, we will add noinline attribute. Is this OK with you?

File: system-contracts/contracts/EvmInterpreterFunctions.template.yul [hunk 1]
```diff
@@ -516,21 +541,12 @@ function expandMemory(newSize) -> gasCost {
 
 // Essentially a NOP that will not get optimized away by the compiler
 function $llvm_NoInline_llvm$_unoptimized() {
-    pop(1)
 }
 
 function printHex(value) {
-    mstore(add(DEBUG_SLOT_OFFSET(), 0x20), 0x00debdebdebdebdebdebdebdebdebdebdebdebdebdebdebdebdebdebdebdebde)
-    mstore(add(DEBUG_SLOT_OFFSET(), 0x40), value)
-    mstore(DEBUG_SLOT_OFFSET(), 0x4A15830341869CAA1E99840C97043A1EA15D2444DA366EFFF5C43B4BEF299681)
```
[the remark this answers is not in the document]
Comment (vladimirradosavljevic) [reply system-contracts/contracts/EvmInterpreterFunctions.template.yul]: I reverted this change, but please note that these checks actually changes LLVM decisions for some opcodes, so I wanted to check the numbers without them.

**Comment by vladimirradosavljevic** [comment]
@jrchatruc Thanks for the review!
Please note that I added one more commit to add always inline to warmAddress function. I saw that in one place it is not inlined, so I added always inline attribute to force inlining in all places.

**Comment by jrchatruc** [comment]
> > > @jrchatruc PTAL. Improvements can be seen [here](https://github.com/matter-labs/era-compiler-tester/pull/62) (please note that this PR contains one pending LLVM optimization that is mostly related to ADD opcode, so we can get 44.750 instead of 46.750).
> > 
> > 
> > This looks really good, wouldn't it be worth it to also force the no inline on the constructor `simulate` function?
> 
> Currently in the constructor `simulate` function that is inlined, LLVM doesn't generate unneeded spills and reloads. In case we see that this happens in a future, we will add noinline attribute. Is this OK with you?

Yep, totally ok, was just curious about it

**Review by jrchatruc — APPROVED** [review]

**Pull request merged**